### PR TITLE
Universal protobuf dump format

### PR DESCRIPTION
**Pull request by BAndysc**
### Problem:
WPP is doing great job in parsing packets, but I think everyone is aware that processing sniff while parsing is not a good idea. Currently there is a lot of processing code in packet handlers, e.g: https://github.com/TrinityCore/WowPacketParser/blob/master/WowPacketParserModule.V6_0_2_19033/Parsers/ChatHandler.cs#L123
This is hard to maintain and change, it is hard to do more more advanced processing, you cannot share code between different version packet handler. **Parsing and processing stage should be separated.**

Parsing .txt output is also not an option, it is too error prone + sometimes even if packet is the same between different wow versions, WPP output text changes. Just no point in parsing .txt output.

### Solution:
The idea is to introduce additional dump format - intermediate format between raw packets and txt output. My idea is to prepare "universal protobuf structures for most common packets", parser will produce such protobuf and it can be later easily edited.

Packets changes over version. Those "universal packets" will try to minimize those differences when possible. It is not intended to keep all packet data, only most common ones. This does NOT replace .txt output.

Why protobuf? Because it has bindings for almost all languages in the world.

This is example way to process parsed sniff, in this example - parsing `creature_text`: https://github.com/BAndysc/WoWDatabaseEditor/blob/packet_viewer/WDE.PacketViewer/Processing/Processors/CreatureTextProcessor.cs

### What is done:
I've added new dump format + basic structures. I've also prepared first specific packet structure - SMSG_CHAT and implemented parsing it in 6.0.x handler.

#### Before I continue my work, I would like to hear your opinion on this and whether it could be merged once done.

### CODE REVIEW

**Comment by mdX7** [comment]
General idea sounds good and then the ugly .txt parsing would be gone in a few tools, finally.

But I kinda have a bad feeling about having redundant data in memory during parsing time when thinking of massparses since im already running out of memory quite fast, this is just a special case though. Maybe we should use some shared structure for wpp and pbuf uses during parsing time and generate the pbuf stuff post parsing? 
Might be quite hard to link it with the packet header data in that case though.
During normal parses this shouldnt be a problem at all, even with big sniffs.
If we are clearing the cached pbuf stuff after parsing is done we should be fine too. (You might be already doing that, didnt read the code completely yet)

Wouldnt go that far to split parsing and processing in general in wpp but a separate dump type is completely fine.

**Comment by BAndysc** [comment]
Indeed, I am not clearing anything now + protobuf structures are filled with data even in other modes than 'proto dump'.

I didn't have any idea have any other idea how to code this. 

> Maybe we should use some shared structure for wpp and pbuf uses during parsing time and generate the pbuf stuff post parsing?

This kinda could work, but if I understand correctly, it would require way too much work imo. I mean: those protobuf structures are made to be "universal" between wow versions*, while such "intermediate" form, would have to precisely match packet in handler, which means it would require creating classes for each packet/version with exact fields as in real packet; sounds like too much work.

 What can be done, is to NULL proto structure reference after handling each packet (in all dump formats but proto) so that garbage collector could collect those when there is need.


* unless it is not possible, then probably I will create two protobuf structures for the same packet in different version

**Comment by jackpoz** [comment]
> im already running out of memory quite fast

In the past we have fixed some cpu and memory bottlenecks as they arised, changing the way the code works (like in https://github.com/TrinityCore/WowPacketParser/commit/486cc07bceb64b759cafdea762ebc3bff85f083a , where we went from "insert and then distinct" to "distinct on every insert", reducing a lot the memory usage).

If you have a simple repro case I could take a look into it

File: WowPacketParser/Loading/SniffFile.cs [hunk 1]
```diff
@@ -248,12 +256,19 @@ private void ProcessFileImpl()
 
                             // Close Writer, Stream - Dispose
                             packet.ClosePacket();
+                            packets.Packets_.Add(packet.Holder);
```
Comment (jackpoz) [inline WowPacketParser/Loading/SniffFile.cs:259]: should this be guarded by a if check so it gets filled only when needed, i.e. when _dumpFormat == DumpFormatType.UniversalProto ?

Comment (BAndysc) [reply WowPacketParser/Loading/SniffFile.cs:259]: Right, only in that one case there is point in adding the packet to the list

Comment (jackpoz) [reply WowPacketParser/Loading/SniffFile.cs:259]: you could set a bool variable "storePackets" or so with the _dumpFormat == DumpFormatType.UniversalProto check so in the future it's easy to change the logic only in that place at the beginning

**Comment by mdX7** [comment]
> 
> 
> > im already running out of memory quite fast
> 
> In the past we have fixed some cpu and memory bottlenecks as they arised, changing the way the code works (like in [486cc07](https://github.com/TrinityCore/WowPacketParser/commit/486cc07bceb64b759cafdea762ebc3bff85f083a) , where we went from "insert and then distinct" to "distinct on every insert", reducing a lot the memory usage).
> 
> If you have a simple repro case I could take a look into it

I don't remember which sql option it actually was which actually let me ran out of memory, but in general the memory usage gets quite high when doing:
Parse like 200 GB of sniffs, filter it to only use relevant opcodes (e.g. `SMSG_GOSSIP_MESSAGE`) and watch memory usage rise after a few mins or hours. Happens to me on Windows and Linux (via Docker)

**Comment by funjoker** [comment]
BTW: is the adding of data to the protobuff store disabled if config is disabled? we don't want to spend twice as much memory for it if we dont need it ^^

**Comment by BAndysc** [comment]
> BTW: is the adding of data to the protobuff store disabled if config is disabled? we don't want to spend twice as much memory for it if we dont need it ^^

Yes, currently only in new UniversalProto/UniversalProtoWithText modes packets are stored in the store: https://github.com/TrinityCore/WowPacketParser/pull/604/files#diff-2e10af2943210837251e6bf37d98b779e3b36ed5c97cc5f8c53b9c601aa85cbbR266

**Comment by BAndysc** [comment]
@jackpoz @funjoker @mdX7 

So it looks like the PR is ready to be reviewed and hopefully merged.

## How does it work
Added new dump formats: `UniversalProto` and `UniversalProtoWithText`. Those two dumpers will write file in protobuf format according to this message definition: https://github.com/TrinityCore/WowPacketParser/blob/ba09008f21768993383d78237b57120756c13cc6/WowPacketParser/PacketStructures/structures.proto.

Not all packets have their protobuf structure, in fact, this is not the point of this change. Only most common, most necessary opcodes have their "universal representation". Of course if there is a need, it is very easy to add new messages.

You can later open such parsed sniff in any app which will use this structure. You can now test it in WoWDatabaseEditor, in [packet_viewer branch](https://ci.appveyor.com/api/projects/BAndysc/wowdatabaseeditor/artifacts/WoWDatabaseEditorWindows.zip?branch=packet_viewer). The editor can open a sniff, has advanced filtering features, in the future I will add more "processors" that will process the sniff in various reasons.

image

The PR mainly adds new things, but because of the way the parser works, it touches a lot of handlers. Almost all added messages are supported by all handlers. The only exception is UPDATE_OBJECT for builds >= 8.1.0. It requires some refactor of code, therefore I prefer to do it in another PR.

Any comments, question?

**Comment by jackpoz** [comment]
If it works, merge it after CI checks pass 😃

File: WowPacketParser.Proto/PacketStructures/structures.proto [hunk 2]
```diff
@@ -204,6 +204,13 @@ message Vec3 {
   float z = 3;
 }
 
+message Vec4 {
+  float x = 1;
+  float y = 2;
+  float z = 3;
+  float w = 4;
```
Comment (Shauren) [inline WowPacketParser.Proto/PacketStructures/structures.proto:211]: this should be `o`, `w` is for quats

**Comment by jackpoz** [comment]
Thanks for the PR :)

**Pull request merged**